Re-implementation of the pyrenews basicrenewlmodel using the proposed modularization

gvegayoncdc opened the pull request:
This PR corresponds to the materialization of the programming pattern proposal. The main changes are:

## Structure

- It creates submodules to organize the code better, namely: `processes`, `observations`, `latent`, and `models`.
- It creates the meta-classes `RandomProcess` and `Model`, which are used by the new submodules. The `Model` class includes the following pre-defined methods:
    - `_init_model()`: Creates the NUTS kernel and MCMC sampler
    - `run()`: runs the MCMC.
    - `print_summary()`: A wrapper of `MCMC.print_summary`.
    - `spread_draws()`: A wrapper of `mcmcutils.spread_draws()` (see below).
- `RandomProcess` and `Model` have a `sample()` function with three parameters: `self`, `random_variables` (dict), and `constants` (dict).
- All instances of `RandomProcess` and `Model` return a `collections.namedtuple` object.

## Re-factoring and organization

- It refactors the `basic.py` model into two different models: `BasicRenewalModel` and `HospitalizationsModel`, the latter composed by `BasicRenewalModel`.
- It moves the function `spread_draws` (from the `pyrenew_demo.ipynp`) to `mcmcutils.py` within the library.

## New functionalities + code

- It creates two new `RandomProcess`' (`latent`): `Infections` and `Hospitalizations` (abstractions from the `basic.py`) model.
- It creates the process `RtRandomWalkProcess` (also from `basic.py`).
- It adds six new test files (which include 9 tests) for each one of the observation processes and model.
- Adds two notebooks: `getting-started` and `pyrenew_demo` (from the previous version of the project)

## Rough edges

- The documentation needs work. Although all parameters and main methods are documented. We need to make a editorial review to make sure it is clear.
- Random tests are WIP. Since we still need to figure out how to properly test random functions (see #4,) the implemented tests make some simple checks, including: The function runs, the shape is as expected, etc.

gvegayoncdc commented:
OK, @dylanhmorris. This is ready for review! Please see the notes at the description of the PR.

dylanhmorris reviewed (changes requested):
Some comments and suggestions inline.

Additional general comment: Right now, we use the name `<Quantity>Observation` for classes that represent a key quantity and _optionally_ allow us to observe that quantity with noise: e.g.`InfectionsObservation` for infections; `HospitalizationsObservation` for hospital admissions. I think this could be misleading, since we also use those classes to sample those quantities when they are latent. Some options:

### Option 1: Base class `Observable`, with optional user-defined observation models
Have these quantities be of base class `Observable` or `ObservableQuantity` and name them things like
```python 
class Infections(Observable)
     ...
class HospitalAdmissions(Observable)
     ...
```
`Observable` classes are made actually observed giving them a non-null observation model at instantiation. This is close the to current pattern, as these lines of `getting-started.ipynb`:

```python
infections_obs = InfectionsObservation(
    gen_int=jnp.array([0.25, 0.25, 0.25, 0.25]),
    inf_observation_model=PoissonObservation(
        rate_varname='infections_mean',
        counts_varname='infections_obs',
        )
    )
```

### Option 2: Inheritance
You could also do this via inheritance, though it's not my preference
```python
class Infections(Observable):
    ...

class ObservedInfections(Infections)
    def observe(self, data):
         ...
```

### Option 3: Fully separate the latent observables from their observation processes
With this option, you'd use an object of class `Infections` to sample predicted infections and then (optionally) an object of class `Observation` (e.g. `NegativeBinomialObservation`, `PoissonObservation`) to observe it, but you'd do that at the level of the `Model` class `model()` function (or `sample()` function if we rename it per my suggestion above).

e.g. something like the following:

```python
class MyModel(Model):
    def __init__(self):
        self.infections = Infections()
        self.obs_dist = NegativeBinomialObservation()
   def model(self, data):
        pred_infections = self.infections.sample()
        inf_obs = self.obs_dist.sample(
             "observed_infections",
             pred_infections)
```
Note that (especially if we go with option three) we go this route I'm not convinced that the base class for things like `Infections`, `HospitalAdmissions` etc should be named `Observable`. We can separately discuss the name.

gvegayoncdc commented on model/src/pyrenew/models/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from pyrenew.models.basicrenewal import BasicRenewalModel  # noqa: F401
+from pyrenew.models.hospitalizations import HospitalizationsModel  # noqa: F401
```
Since the imported components are not used in `__init__.py`, pre-commit complains! The same happens in numpyro (or jax?,) if I remember correctly.

dylanhmorris replied:
In general, the recommended practice here is to use `__all__` https://stackoverflow.com/a/49266468

I wonder why linters don't suggest this when they find unused imports specifically in an `__init__.py`

gvegayoncdc replied:
I was not sure if using `__all__` was something we wanted, but I'm happy to follow that PEP!

dylanhmorris replied:
Sounds good. I think we should.

gvegayoncdc commented on model/src/pyrenew/mcmcutils.py:
```diff
@@ -0,0 +1,94 @@
+
+
+def spread_draws(posteriors, variable_names):
+    """
```
Yes, I hear you. Documenting with sphinx style is not pretty. I'll change all of those.

gvegayoncdc commented on model/src/pyrenew/metaclasses.py:
```diff
@@ -0,0 +1,169 @@
+
+    @abstractmethod
+    def model(self, data):
+        pass
```
Agreed, I was thinking about this. My approach in R would be to create an alias, but it could be overkill. I'll go with `sample()` instead.

gvegayoncdc commented on model/src/pyrenew/observations/__init__.py:
```diff
@@ -0,0 +1,13 @@
+    sample_infections_rt,
+    sample_infections_with_feedback,
+)
+from pyrenew.observations.infections import InfectionsObservation  # noqa: F401
```
Same as before, since the thing is not used in `__init__.py`, precommit complains.

gvegayoncdc commented on model/src/pyrenew/metaclasses.py:
```diff
@@ -0,0 +1,169 @@
+        """
+        return self.mcmc.print_summary(prob, exclude_deterministic)
+
+    def plot(self):
+        pass
```
I was thinking a basic plot would be to show the distribution/trajectory of the first `k` sampled variables.

gvegayoncdc replied:
I've removed it for now.

gvegayoncdc commented on model/src/pyrenew/observations/negativebinomial.py:
```diff
@@ -0,0 +1,97 @@
+            imply that the distribution becomes
+            more Poissonian, while smaller ones
+            imply a greater degree of dispersion.
+        :type concentration_prior: dist.Distribution
```
The validate function checks for that. I've extended the documentation.

gvegayoncdc commented:
OK @dylanhmorris, the latest version features the following changes:

1. The `__init__.py` script now uses `__all__` to avoid the pre-commit comment.
2. `HospitalizationsModel` now uses `BasicRenewalModel` as a component, no inheritance.
3. All documentation now uses the `numpy` style.
4. Given your comments, @SamuelBrand1, and @seabbs regarding latent vs observed, we now follow approach 3 in your comment, i.e., latent and observed process are fully split.
5. Removed the plot method from Model (for now)
6. Removed the spread indices function from mcmcutils.py

We can discuss details tomorrow during the team meeting.

gvegayoncdc commented:
> OK @dylanhmorris, the latest version features the following changes:
> 
> 1. The `__init__.py` script now uses `__all__` to avoid the pre-commit comment.
> 2. `HospitalizationsModel` now uses `BasicRenewalModel` as a component, no inheritance.
> 3. All documentation now uses the `numpy` style.
> 4. Given your comments, @SamuelBrand1, and @seabbs regarding latent vs observed, we now follow approach 3 in your comment, i.e., latent and observed process are fully split.
> 5. Removed the plot method from Model (for now)
> 6. Removed the spread indices function from mcmcutils.py
> 
> We can discuss details tomorrow during the team meeting.

Also:

- The `Hospitalizations` latent class is called `HospitalAdmissions`
- Previously, the `HospitalizationsModel` would receive a RandomProcess for observed infections. This is unlikely to happen and confusing so I've removed it.

dylanhmorris commented on model/src/pyrenew/processes/rtrandomwalk.py:
```diff
@@ -0,0 +1,89 @@
+from pyrenew.transform import AbstractTransform, LogTransform
+
+
+class RtRandomWalkProcess(RandomProcess):
```
Given that we now have a `latent` module for latent (but potentially observable) variables, I propose creating an an `Rt` class there. 

To represnt $\mathcal{R}(t)$ varying in time according to a particular stochastic process on a particular transformed scale, one would instantiate the `Rt` class (or a subclass of it), using stochastic processes from `pyrenew.processes`, transformations from `pyrenew.transform` etc.

dylanhmorris replied:
Works for me. Create an issue?

dylanhmorris reviewed (commented):
We should think about module naming on the following fronts:

1. Plural versus singular
2. Noun-like versus adjective-like

Currently we have `pyrenew.processes` and `pyrenew.observations` (plural nouns) and `pyrenew.latent` (adjective)

dylanhmorris commented on model/src/pyrenew/metaclasses.py:
```diff
@@ -0,0 +1,251 @@
+    return None
+
+
+class RandomProcess(metaclass=ABCMeta):
```
I think given the new architecture we may want to rename this simply `RandomVariable`; we also need to rewrite the docstring given how we're now using it.

gvegayoncdc commented:
Thank you, @dylanhmorris! That was 💨!
